File: x/agentplane/runner/harness_background_child.c
#include <stdio.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>

/* A native-harness stand-in for verifying inherited state ownership across a leader exit. */
int main(void) {
  pid_t child = fork();
  if (child < 0) {
    perror("fork background harness child");
    return 125;
  }
  if (child == 0) {
    signal(SIGTERM, SIG_IGN);
    for (;;) {
      pause();
    }
  }
  printf("%ld\n", (long)child);
  fflush(stdout);
  for (;;) {
    pause();
  }
}

File: agentplane/runner/harness_background_child.rs
use std::io::{self, Write};

use nix::sys::signal::{SigHandler, Signal, signal};
use nix::unistd::{ForkResult, fork, pause};

// Native-harness stand-in verifying inherited ownership across leader exit.
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // SAFETY: this process is single-threaded. The child only calls pause after fork, and signal
    // dispositions use constants, not Rust callbacks. Set SIG_IGN before fork so the published
    // child PID also proves it is ready to ignore SIGTERM; the leader restores default handling.
    unsafe {
        signal(Signal::SIGTERM, SigHandler::SigIgn)?;
        match fork()? {
            ForkResult::Child => loop {
                pause();
            },
            ForkResult::Parent { child } => {
                signal(Signal::SIGTERM, SigHandler::SigDfl)?;
                println!("{child}");
                io::stdout().flush()?;
            }
        }
    }
    loop {
        pause();
    }
}
